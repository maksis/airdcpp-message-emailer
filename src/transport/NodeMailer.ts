import nodemailer from 'nodemailer';

import { Context, SeverityEnum } from '../types';
import { hasConfig } from '../utils';

export const NodeMailer = (context: Context) => {
  const { api, logger, getExtSetting, extension } = context;

  let hostname: string;
  let transporter: nodemailer.Transporter | undefined;

  const constructSmtpSettings = () => {
    const ret = {
      host: getExtSetting('smtp_host'),
      port: getExtSetting('smtp_port'),
      secure: getExtSetting('smtp_secure'),
    };

    const user = getExtSetting('smtp_user');
    const pass = getExtSetting('smtp_password');
    if (user && pass) {
      ret['auth'] = {
        user,
        pass,
      };
    }

    return ret;
  };

  const constructMail = (text: string) => {
    const mailOptions: nodemailer.SendMailOptions = {
      from: getExtSetting('from'),
      to: getExtSetting('to'),
      subject: `Chat summary from AirDC++ (${hostname})`,
      text,
    };

    return mailOptions;
  };

  const createTransporter = () => {
    const smtpSettings = constructSmtpSettings();
    return nodemailer.createTransport(smtpSettings);
  };

  const sendMail = (messageSummary: string) => {
    transporter!.sendMail(constructMail(messageSummary), (error: any, info: any) => {
      if (error) {
        logger.error(`Failed to send email: ${error}`);
        api.postEvent(`${extension.name}: failed to email message summary (see the extension error log for more information)`, SeverityEnum.ERROR);
      } else {
        logger.info(`Message ${info.messageId} sent: ${info.response}`);
      }
    });
  }

  
  const reset = (newHostName: string) => {
    hostname = newHostName;

    logger.verbose('Creating mailer interface');
    try {
      transporter = nodemailer.createTransport(constructSmtpSettings());
    } catch (e) {
      logger.error(`Failed to initialize mailer interface: ${e}`);
      process.exit(1);
    }

    if (hasConfig(getExtSetting)) {
      transporter.verify((error: any, _success: boolean) => {
        logger.verbose('Verifying settings');
        if (error) {
          api.postEvent(`${extension.name}: SMTP setting validation failed (${error})`, SeverityEnum.ERROR);
        } else {
          logger.verbose('Setting validation succeeded');
        }
      });
    }
  }

  return {
    createTransporter,
    sendMail,
    reset,
  };
};

export type Transporter = ReturnType<typeof NodeMailer>;
