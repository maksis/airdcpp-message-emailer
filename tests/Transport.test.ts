import { describe, test, expect, vi } from 'vitest';

import { Mailer, NodeMailer } from '../src/transport/NodeMailer';
import { createMockContext } from './mocks/mock-context';


describe('Transport', () => {

  test('should send email', async () => {
    const transport: Mailer = {
      sendMail: vi.fn(),
      verify: vi.fn(),
    };

    const createTransport = vi.fn(() => transport);

    const context = createMockContext();
    const { sendMail, reset } = NodeMailer(context, createTransport);

    reset('mock hostname');

    expect(transport.verify).toHaveBeenCalled();

    sendMail('Test message summary');
    expect(transport.sendMail).toMatchSnapshot();
  });
});
