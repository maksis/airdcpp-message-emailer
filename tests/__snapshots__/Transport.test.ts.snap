// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transport > should send email 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "from": ""AirDC++" <example@example.com>",
        "subject": "Chat summary from AirDC++ (mock hostname)",
        "text": "Test message summary",
        "to": "to@example.com",
      },
      [Function],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
